File: data-structures/selection-sort/selection-sort.java

import java.util.Arrays;

//code in Java
public class Main {
  public static void selectionSort(int [] numbers) {
    int i,j,indexSmallest,temp;

    //##### selection sort loop
    //this incrementally moves the boundary of the unsorted array
    for (i = 0; i < numbers.length - 1; ++i) {
      // define smallest index as first index value initially
      indexSmallest = i;
      // find the index of the smallest element
      for (j = i + 1; j < numbers.length; ++j) {
        //if j is less than i, then j becomes smallest index
        if (numbers[j] < numbers[indexSmallest]){
          //set smallest index as j
          indexSmallest = j;
        }
      }
      //do swap of smallest index when needed
      //save current number at ith index to temp
      temp = numbers[i];
      //place number at indexsmallest index to ith index
      numbers[i] = numbers[indexSmallest];
      //place number at temp into index of smallest index
      numbers[indexSmallest] = temp;
      }
     //##### end of selection sort loop
    
    //swap minimum element and first elements
    //since now fully swapped, minimum element is index 0
    //repeats process of temp swap above, with 0 as the smallest index and 1 as the ith index
    indexSmallest=0;
    temp=numbers[indexSmallest];
    numbers[indexSmallest] = numbers[1];
    numbers[1] = temp;
  }

public static void main(String[] args) {
  //create array
  int[] array = new int[21];
  // set up first values for array
  int numval=50;
  int i=0;
  //fill array
  for ( i=0; i<array.length; i++){
    array[i]=numval;
    numval++;
  }
  //print out initial array
  System.out.println("INITIAL ARRAY: ");
  System.out.println(Arrays.toString(array));
  
  //apply selection sort (with the minimum and first values switched)
  selectionSort(array);
  System.out.println("SELECTION SORTED: ");
  System.out.println(Arrays.toString(array));
  }
}

File: data-structures/selection-sort/selection-sort.py
#CIS 131 Exam 1
#Rosemary Moseley
#Due: 9/14/23


#Selection Sort in Python
def selectionSort(array, size):
  #iterating over all items in array
  for i in range(size):
    #define starting smallest index as first element
    indexSmallest = i
    #check next elemement to see if it is smaller than the previous element
    for j in range(i + 1, size):
      # select the minimum element in every iteration
      if array[j] < array[indexSmallest]:
        #if next element is smaller, make that the new smallest element
        indexSmallest = j
  # swapping the elements to sort the array
  (array[i], array[indexSmallest]) = (array[indexSmallest], array[i])
  #switch 0 and 1 arrays at the end
  (array[0], array[1]) = (array[1], array[0])


#create empty array
ifill = 50
arr = []
#fill empty array
for i in range(0, 21):
  arr.append(ifill)
  ifill = ifill + 1

size = len(arr)
print('Before Selection Sort is:')
print(arr)
selectionSort(arr, size)
print('Array Afer Selection Sort is:')
print(arr)
